test/Datadog.Trace.Tests/DelegatesHelperTests.cs

```diff
@@ -0,0 +1,137 @@
+            TargetObject.MyEvent -= TargetObject_MyEvent2;
+
+            Assert.Equal(1, lstEvents[0]);
+            Assert.Equal(2, lstEvents[1]);
+            Assert.Equal(3, lstEvents[2]);
+
+            void TargetObject_MyEvent(object sender, EventArgs e)
+            {
+                lstEvents.Add(1);
+
+                var targetDelegate = TargetObject.GetDelegate();
+
+                var lastDelegate = new EventHandler((s, eArgs) =>
+                {
+                    lstEvents.Add(3);
+                });
+
+                if (!DelegatesHelper.TrySetLastDelegate(targetDelegate, lastDelegate))
+                {
+                    lastDelegate(sender, e);
+                }
+            }
+
+            void TargetObject_MyEvent2(object sender, EventArgs e) => lstEvents.Add(2);
+        }
+
+        [Fact]
+        public void SetLastDelegateInversedTest()
+        {
+            var lstEvents = new List<int>();
+
+            TargetObject.MyEvent += TargetObject_MyEvent2;
+            TargetObject.MyEvent += TargetObject_MyEvent;
+            TargetObject.FireEvent();
+            TargetObject.MyEvent -= TargetObject_MyEvent2;
+            TargetObject.MyEvent -= TargetObject_MyEvent;
+
+            Assert.Equal(2, lstEvents[0]);
+            Assert.Equal(1, lstEvents[1]);
+            Assert.Equal(3, lstEvents[2]);
+
+            void TargetObject_MyEvent(object sender, EventArgs e)
+            {
+                lstEvents.Add(1);
+
+                var targetDelegate = TargetObject.GetDelegate();
+
+                var lastDelegate = new EventHandler((s, eArgs) =>
+                {
+                    lstEvents.Add(3);
+                });
+
+                if (!DelegatesHelper.TrySetLastDelegate(targetDelegate, lastDelegate))
+                {
+                    lastDelegate(sender, e);
+                }
+            }
+
+            void TargetObject_MyEvent2(object sender, EventArgs e) => lstEvents.Add(2);
+        }
+
+        [Fact]
+        public void SetLastDelegaten3HandlersTest()
+        {
+            var lstEvents = new List<int>();
+
+            TargetObject.MyEvent += TargetObject_MyEvent2;
+            TargetObject.MyEvent += TargetObject_MyEvent;
+            TargetObject.MyEvent += TargetObject_MyEvent2;
+            TargetObject.FireEvent();
+            TargetObject.MyEvent -= TargetObject_MyEvent2;
+            TargetObject.MyEvent -= TargetObject_MyEvent;
+            TargetObject.MyEvent -= TargetObject_MyEvent2;
+
+            Assert.Equal(2, lstEvents[0]);
+            Assert.Equal(1, lstEvents[1]);
+            Assert.Equal(2, lstEvents[2]);
+            Assert.Equal(3, lstEvents[3]);
+
+            void TargetObject_MyEvent(object sender, EventArgs e)
+            {
+                lstEvents.Add(1);
+
+                var targetDelegate = TargetObject.GetDelegate();
+
+                var lastDelegate = new EventHandler((s, eArgs) =>
```

Comment:
In a separate unit test, can you set `var lastDelegate` to `TargetObject_MyEvent` to make sure it doesn't add itself? Based on my understanding, that's the role of the small stackframe capture check in `TrySetLastDelegate`

Reply:
done in f861559